controller/project/manager.py

```diff
@@ -219,12 +229,182 @@ def __get_crowd_label_is_data(project_id: str, is_id: str) -> Dict[str, str]:
     if huddle_type == enums.LinkTypes.DATA_SLICE.value:
-        slices = ds_manager.get_all(project_id, enums.SliceTypes.STATIC_DEFAULT)
+        slices = data_slice.get_all(project_id, enums.SliceTypes.STATIC_DEFAULT)
         if slices and len(slices) > 0:
             return slices[0].id
     elif huddle_type == enums.LinkTypes.HEURISTIC.value:
-        return information_source_manager.get_first_crowd_is_for_annotator(
-            project_id, user_id
-        )
+        return information_source.get_first_crowd_is_for_annotator(project_id, user_id)
     else:
         raise ValueError("invalid huddle type")
+
+
+# GATES
+projects_updating = set()
+projects_updating_lock = threading.Lock()
+
+
+def get_gates_integration_data(project_id: str) -> GatesIntegrationData:
+
+    project_item = project.get(project_id)
+    if not project_item:
+        raise GraphQLError("Project not found")
+
+    if project_id in projects_updating:
+        return GatesIntegrationData(status=enums.GatesIntegrationStatus.UPDATING.value)
+
+    missing_tokenizer = not __tokenizer_pickle_exists(project_item.tokenizer)
+    missing_embeddings = __get_missing_embedding_pickles(project_id)
+    missing_information_sources = __get_missing_information_source_pickles(project_id)
+
+    status = enums.GatesIntegrationStatus.READY.value
+    if (
+        missing_tokenizer
+        or len(missing_embeddings) > 0
+        or len(missing_information_sources) > 0
+    ):
+        status = enums.GatesIntegrationStatus.NOT_READY.value
+
+    return GatesIntegrationData(
+        status=status,
+        missing_tokenizer=missing_tokenizer,
+        missing_embeddings=missing_embeddings,
+        missing_information_sources=missing_information_sources,
+    )
+
+
+def __tokenizer_pickle_exists(config_string: str) -> bool:
+    tokenizer_path = os.path.join(
+        "/inference/tokenizers", f"tokenizer-{config_string}.pkl"
+    )
+    return os.path.exists(tokenizer_path)
+
+
+def __get_missing_embedding_pickles(project_id: str) -> List[str]:
+    missing = []
+    embedding_items = embedding.get_finished_embeddings(project_id)
+    for embedding_item in embedding_items:
+        embedding_id = str(embedding_item.id)
+        emb_path = os.path.join(
+            "/inference", project_id, f"embedder-{embedding_id}.pkl"
+        )
+        if not os.path.exists(emb_path):
+            missing.append(embedding_id)
+    return missing
+
+
+def __get_missing_information_source_pickles(project_id: str) -> List[str]:
+    missing = []
+    is_items = information_source.get_all(project_id)
+    for is_item in is_items:
+        if is_item.type != enums.InformationSourceType.ACTIVE_LEARNING.value:
+            # only active learning information sources are pickled
+            continue
+        is_id = str(is_item.id)
+        last_payload = information_source.get_last_payload(project_id, is_id)
+        if last_payload.state == enums.PayloadState.FINISHED.value:
```

Comment:
if the last one is e.g. the 23rd iteration (and a previous worked) we would still have it or do we delete the al model on error?

Also I'm not sure this is what we intend here. Assuming we only have one active learner that only had error cases so far we would still consider the project gates ready, right? (Findings discussion point)

Reply:
added to findings